rdma-ndd/rdma-ndd.8.in
.\" Man page generated from reStructuredText.
.
.TH RDMA-NDD 8 "@BUILD_DATE@" "" "OpenIB Diagnostics"
.SH NAME
RDMA-NDD \- RDMA device Node Description update daemon
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
rdma\-ndd <options>
.SH DESCRIPTION
.sp
rdma\-ndd is a system daemon which watches for rdma device changes and/or
hostname changes and updates the Node Description of the rdma devices based on
those changes.
.SH DETAILS
.sp
Full operation of this daemon requires kernels which support polling of the
procfs hostname file as well as libudev.
.sp
If your system does not support either of these features, the daemon will set
the Node Descriptions at start up and then sleep forever.
.SS Node Description configuration
.sp
The daemon uses the environment variable RDMA_NDD_ND_FORMAT to set the node
description.  The following wild cards can be specified for more dynamic
control.
.sp
%h \-\- replace with the current hostname (not including domain)
.sp
%d \-\- replace with the device name (for example mlx4_0, qib0, etc.)
.sp
If not specified the default is "%h %d".
.sp
NOTE: At startup, and on new device detection, the Node Description is always
written to ensure the SM and rdma\-ndd are in sync.  Subsequent events will only
write the Node Description on a device if it has changed.
.SS Using systemd
.sp
Setting the environment variable for the daemon is normally be done via a
systemd drop in unit.  For example the following could be added to a file named
/etc/systemd/system/rdma\-ndd.service.d/nd\-format.conf to use only the
hostname as your node description.
.sp
[Service]
Environment="RDMA_NDD_ND_FORMAT=%%h"
.sp
NOTE: Systemd requires an extra \(aq%\(aq.
.SH OPTIONS
.sp
\fB\-f, \-\-foreground\fP
Run in the foreground instead of as a daemon
.SH AUTHOR
.INDENT 0.0
.TP
.B Ira Weiny
< \fI\%ira.weiny@intel.com\fP >
.UNINDENT
.\" Generated by docutils manpage writer.
.


rdma-ndd/rdma-ndd.c
/*
 * Copyright (c) 2014,2016 Intel Corporation. All Rights Reserved
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

#include <config.h>

#include <poll.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <assert.h>
#include <string.h>
#include <limits.h>
#include <stdio.h>
#include <syslog.h>
#include <dirent.h>
#include <errno.h>
#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>

#include <libudev.h>

static struct udev *g_udev;
static struct udev_monitor *g_mon;

#define SYS_HOSTNAME "/proc/sys/kernel/hostname"
#define SYS_INFINIBAND "/sys/class/infiniband"
#define DEFAULT_ND_FORMAT "%h %d"

static char *g_nd_format = NULL;

static void newline_to_null(char *str)
{
	char *term = index(str, '\n');
	if (term)
		*term = '\0';
}

static void strip_domain(char *str)
{
	char *term = index(str, '.');
	if (term)
		*term = '\0';
}

static void build_node_desc(char *dest, size_t len,
		     const char *device, const char *hostname)
{
	char *end = dest + len-1;
	const char *field;
	char *src = g_nd_format;

	while (*src && (dest < end)) {
		if (*src != '%') {
			*dest++ = *src++;
		} else {
			src++;
			switch (*src) {
			case 'h':
				field = hostname;
				while (*field && (*field != '.') && (dest < end))
					*dest++ = *field++;
				break;
			case 'd':
				field = device;
				while (*field && (dest < end))
					*dest++ = *field++;
				break;
			}
			src++;
		}
	}
	*dest = 0;
}

static int update_node_desc(const char *device, const char *hostname, int force)
{
	int rc;
	char nd[128];
	char new_nd[64];
	char nd_file[PATH_MAX];
	FILE *f;

	snprintf(nd_file, sizeof(nd_file), SYS_INFINIBAND "/%s/node_desc",
			device);
	nd_file[sizeof(nd_file)-1] = '\0';

	f = fopen(nd_file, "r+");
	if (!f) {
		syslog(LOG_ERR, "Failed to open %s\n", nd_file);
		return -EIO;
	}

	if (!fgets(nd, sizeof(nd), f)) {
		syslog(LOG_ERR, "Failed to read %s\n", nd_file);
		rc = -EIO;
		goto error;
	}
	newline_to_null(nd);

	build_node_desc(new_nd, sizeof(new_nd), device, hostname);

	if (!force && strncmp(new_nd, nd, sizeof(new_nd)) == 0) {
		syslog(LOG_DEBUG, "%s: no change (%s)\n", device, new_nd);
	} else {
		syslog(LOG_DEBUG, "%s: change (%s) -> (%s)\n",
			device, nd, new_nd);
		rewind(f);
		fprintf(f, "%s", new_nd);
	}

	rc = 0;
error:
	fclose(f);
	return rc;
}

static void set_rdma_node_desc(const char *hostname, int force)
{
	DIR *class_dir;
	struct dirent *dent;

	class_dir = opendir(SYS_INFINIBAND);
	if (!class_dir) {
		syslog(LOG_ERR, "Failed to open " SYS_INFINIBAND);
		return;
	}

	while ((dent = readdir(class_dir))) {
		if (dent->d_name[0] == '.')
			continue;

		if (update_node_desc(dent->d_name, hostname, force))
			syslog(LOG_ERR, "set Node Description failed on %s\n",
			       dent->d_name);
	}

	closedir(class_dir);
}

static void read_hostname(int fd, char *name, size_t len)
{
	memset(name, 0, len);
	if (read(fd, name, len-1) >= 0) {
		newline_to_null(name);
		strip_domain(name);
	} else {
		syslog(LOG_ERR, "Read %s Failed\n", SYS_HOSTNAME);
	}
	lseek(fd, 0, SEEK_SET);
}

static void setup_udev(void)
{
	g_udev = udev_new();
	if (!g_udev) {
		syslog(LOG_ERR, "udev_new failed\n");
		return;
	}
}

static int get_udev_fd(void)
{
	g_mon = udev_monitor_new_from_netlink(g_udev, "udev");
	if (!g_mon) {
		syslog(LOG_ERR, "udev monitoring failed\n");
		return -1;
	}

	udev_monitor_filter_add_match_subsystem_devtype(g_mon, "infiniband", NULL);
	udev_monitor_enable_receiving(g_mon);
	return udev_monitor_get_fd(g_mon);
}

static void process_udev_event(int ud_fd, const char *hostname)
{
	struct udev_device *dev;

	dev = udev_monitor_receive_device(g_mon);
	if (dev) {
		const char *device = udev_device_get_sysname(dev);
		const char *action = udev_device_get_action(dev);

		syslog(LOG_DEBUG, "Device event: %s, %s, %s\n",
			udev_device_get_subsystem(dev),
			device, action);

		if (device && action
		    && strncmp(action, "add", sizeof("add")) == 0)
			if (update_node_desc(device, hostname, 1))
				syslog(LOG_ERR, "set Node Description failed on %s\n",
				       device);

		udev_device_unref(dev);
	}
}

static void monitor(void)
{
	char hostname[128];
	int hn_fd;
	struct pollfd fds[2];
	int numfds = 1;
	int ud_fd;

	hn_fd = open(SYS_HOSTNAME, O_RDONLY);
	if (hn_fd < 0) {
		syslog(LOG_ERR, "Open %s Failed exiting\n",
			SYS_HOSTNAME);
		exit(1);
	}

	read_hostname(hn_fd, hostname, sizeof(hostname));
	set_rdma_node_desc((const char *)hostname, 1);

	fds[0].fd = hn_fd;
	fds[0].events = 0;

	ud_fd = get_udev_fd();
	if (ud_fd >= 0)
		numfds = 2;

	fds[1].fd = ud_fd;
	fds[1].events = POLLIN;

	while (1) {
		if (poll(fds, numfds, -1) <= 0) {
			syslog(LOG_ERR, "Poll %s failed; exiting\n", SYS_HOSTNAME);
			exit(1);
		}

		if (fds[0].revents != 0) {
			read_hostname(hn_fd, hostname, sizeof(hostname));
			syslog(LOG_DEBUG, "Hostname event: %s\n", hostname);
			set_rdma_node_desc((const char *)hostname, 0);
		}

		if (fds[1].revents != 0)
			process_udev_event(ud_fd, hostname);
	}
}

int main(int argc, char *argv[])
{
	int foreground = 0;

	openlog("rdma-ndd", LOG_PID | LOG_PERROR, LOG_DAEMON);

	while (1) {
		int opt_idx = 0;
		static struct option long_opts[] = {
			{"foreground",   0, 0, 'f' },
			{"help",         0, 0, 'h' },
			{0,              0, 0,  0 }
		};

		char c = getopt_long(argc, argv, "fh", long_opts, &opt_idx);
		if (c == -1)
			break;

		switch (c) {
		case 'f':
			foreground = 1;
			break;
		case 'h':
			printf("rdma-ndd [options]\n");
			printf("    See 'man rdma-ndd' for details\n");
			exit(0);
			break;
		default:
			break;

		}
	}

	if (!foreground) {
		closelog();
		openlog("rdma-ndd", LOG_PID, LOG_DAEMON);
		if (daemon(0, 0) != 0) {
			syslog(LOG_ERR, "Failed to daemonize\n");
			exit(errno);
		}
	}

	setup_udev();

	g_nd_format = getenv("RDMA_NDD_ND_FORMAT");
	if (g_nd_format && strncmp("", g_nd_format, strlen(g_nd_format)) != 0)
		g_nd_format = strdup(g_nd_format);
	else
		g_nd_format = strdup(DEFAULT_ND_FORMAT);

	syslog(LOG_DEBUG, "Node Descriptor format (%s)\n", g_nd_format);

	monitor();

	return 0;
}
